drivers: can: mchp_mss: New

sebhub opened the pull request:
Add a CAN driver for Microchip's PolarFire SoC platform.

sebhub committed:
drivers: can: driver for Microchip's PolarFire SoC

Add a CAN driver for Microchip's PolarFire SoC platform.

Co-authored-by: Donato Brusamento <donato.brus@gmail.com>
Co-authored-by: Alexander Wachter <alexander@wachter.cloud>
Signed-off-by: Sebastian Huber <sebastian.huber@embedded-brains.de>

henrikbrixandersen reviewed (changes requested):
@sebhub Please address the comments regarding endianness delivered out-of-band in https://github.com/zephyrproject-rtos/zephyr/pull/96938

DonBrus commented:
> @sebhub Please address the comments regarding endianness delivered out-of-band in #96938

@sebhub to expand on this, using the API as described in the official docs, e.g. calling `can_send` with the frame's `data` set to `.data = {1,2,3,4,5,6,7,8}` will result in the respective transmission register being set to
```
TX_MSGx _DATA_HIGH  04030201
TX_MSGx _DATA_LOW   08070605
```
due to type punning in [mss_can_send](https://github.com/zephyrproject-rtos/zephyr/blob/8209b29a6c95a6eec0376f62719f6d40f25c2659/drivers/can/can_mchp_mss.c#L368). The controller sends out bytes MSB to LSB, hence the bus order inversion (`0x04 0x03 0x02 0x01 0x08 0x07 0x06 0x05`).
It's possible to get around this without inverting byte order in software in the frame data by setting the `CAN_CONFIG.SWAP_ENDIAN` bit, as described in MC docs

sebhub commented:
I had a flaw in my test setup. I use an MPFS which is connected through a CAN bus to an STM32H7 running Zephyr. In my initial test setup, the STM32H7 simply retransmitted received CAN messages. This setup cannot detect these endian issues. I changed the test setup to check also transfers in one direction. Indeed, the swap endian configuration is required. I pushed a commit based on #96938 which fixes the issue.

con-pax commented on drivers/can/Kconfig.mchp_mss:
```diff
@@ -0,0 +1,9 @@
+
+config CAN_MCHP_MSS
+	bool "Microchip PolarFire SoC CAN driver"
+	depends on DT_HAS_MICROCHIP_MPFS_CAN_ENABLED
```
small nit: can you add `default y` under `bool "Microchip PolarFire SoC CAN driver"` so that a user only needs to set `CONFIG_CAN=y` in their project please

con-pax reviewed (dismissed):
LGTM

alexanderwachter reviewed (changes requested):
... and the rest that can be const.

con-pax reviewed (dismissed):
LGTM

alexanderwachter reviewed (dismissed):
Thanks @sebhub. First CAN on a RISCV as far as I see 🎉

str4t0m commented on boards/microchip/mpfs_icicle/mpfs_icicle_polarfire_e51_defconfig:
```diff
@@ -13,3 +13,4 @@ CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
 # GPIO driver options
 CONFIG_GPIO=y
 CONFIG_I2C=y
+CONFIG_CAN=y
```
The boards should not enable CAN per default, only applications needing it should enable it.

sebhub replied:
@con-pax Please have a look at this comment.

str4t0m commented on dts/riscv/microchip/mpfs.dtsi:
```diff
@@ -309,5 +309,23 @@
+			reg = <0x2010c000 0x1000>;
+			interrupt-parent = <&plic>;
+			interrupts = <56 1>;
+			clock-frequency = <8000000>;
```
Seeing, that this not used to configure anything in the driver, only to be able to report the frequency. Where are are these configured?

alexanderwachter replied:
80MHz is a CiA recommended clock frequency.

henrikbrixandersen reviewed (changes requested):
@sebhub Please confirm that at least the `tests/drivers/can/timing` and `tests/drivers/can/api` test suites pass with hardware-in-the-loop testing.

Also, please split this into multiple commits (e.g. one commit for adding the DT binding, one for adding the driver, one for adding board support) and use Git commit logs according the scheme used throughout the tree.

str4t0m reviewed (changes requested):
Some minor comments.

sebhub committed:
dts: riscv: microchip: add MPFS CAN support

Add a CAN device tree binding and nodes for Microchip's PolarFire SoC
platform.

Signed-off-by: Sebastian Huber <sebastian.huber@embedded-brains.de>

con-pax committed:
boards: microchip: mpfs_icicle: Add CAN support

Add a CAN support for the Microchip Icicle Kit (MPFS development board).

Co-authored-by: Sebastian Huber <sebastian.huber@embedded-brains.de>
Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

sebhub commented on dts/bindings/can/microchip,mpfs-can.yaml:
```diff
@@ -0,0 +1,38 @@
+  interrupts:
+    required: true
+
+  clock-frequency:
```
Ok, makes sense, however, I have currently only time to add a CAN driver.

@con-pax Could you please comment on the MPFS clock driver.

sebhub commented:
> @sebhub Please confirm that at least the `tests/drivers/can/timing` and `tests/drivers/can/api` test suites pass with hardware-in-the-loop testing.
> 
> Also, please split this into multiple commits (e.g. one commit for adding the DT binding, one for adding the driver, one for adding board support) and use Git commit logs according the scheme used throughout the tree.

I did split up the commits.

I did run the tests with `CONFIG_CAN_MANUAL_RECOVERY_MODE = n`:
```
Running TESTSUITE can_classic
===================================================================
START - test_add_filter
 PASS - test_add_filter in 0.001 seconds
===================================================================
START - test_add_filter_without_callback
 PASS - test_add_filter_without_callback in 0.001 seconds
===================================================================
START - test_add_invalid_ext_filter
E: invalid filter with extended (29-bit) CAN ID 0x1fffffff, CAN ID mask 0x20000000
E: invalid filter with extended (29-bit) CAN ID 0x20000000, CAN ID mask 0x1fffffff
 PASS - test_add_invalid_ext_filter in 0.015 seconds
===================================================================
START - test_add_invalid_null_filter
 PASS - test_add_invalid_null_filter in 0.001 seconds
===================================================================
START - test_add_invalid_std_filter
E: invalid filter with standard (11-bit) CAN ID 0x7ff, CAN ID mask 0x800
E: invalid filter with standard (11-bit) CAN ID 0x800, CAN ID mask 0x7ff
 PASS - test_add_invalid_std_filter in 0.013 seconds
===================================================================
START - test_bitrate_limits
 PASS - test_bitrate_limits in 0.001 seconds
===================================================================
START - test_classic_get_capabilities
 PASS - test_classic_get_capabilities in 0.001 seconds
===================================================================
START - test_filters_added_while_stopped
 PASS - test_filters_added_while_stopped in 0.002 seconds
===================================================================
START - test_filters_preserved_through_bitrate_change
 PASS - test_filters_preserved_through_bitrate_change in 0.003 seconds
===================================================================
START - test_filters_preserved_through_mode_change
 PASS - test_filters_preserved_through_mode_change in 0.003 seconds
===================================================================
START - test_get_core_clock
 PASS - test_get_core_clock in 0.001 seconds
===================================================================
START - test_get_state
 PASS - test_get_state in 0.001 seconds
===================================================================
START - test_invalid_sample_point
 PASS - test_invalid_sample_point in 0.001 seconds
===================================================================
START - test_max_ext_filters
 PASS - test_max_ext_filters in 0.001 seconds
===================================================================
START - test_max_std_filters
 PASS - test_max_std_filters in 0.001 seconds
===================================================================
START - test_receive_timeout
 PASS - test_receive_timeout in 0.100 seconds
===================================================================
START - test_reject_ext_id_rtr
 PASS - test_reject_ext_id_rtr in 0.101 seconds
===================================================================
START - test_reject_std_id_rtr
 PASS - test_reject_std_id_rtr in 0.101 seconds
===================================================================
START - test_send_and_forget
 PASS - test_send_and_forget in 0.001 seconds
===================================================================
START - test_send_callback
 PASS - test_send_callback in 0.001 seconds
===================================================================
START - test_send_ext_id_dlc_of_range
 PASS - test_send_ext_id_dlc_of_range in 0.001 seconds
===================================================================
START - test_send_ext_id_out_of_range
E: invalid frame with extended (29-bit) CAN ID 0x20000000
 PASS - test_send_ext_id_out_of_range in 0.006 seconds
===================================================================
START - test_send_fd_format
 PASS - test_send_fd_format in 0.001 seconds
===================================================================
START - test_send_invalid_dlc
 PASS - test_send_invalid_dlc in 0.001 seconds
===================================================================
START - test_send_null_frame
 PASS - test_send_null_frame in 0.001 seconds
===================================================================
START - test_send_receive_ext_id
 PASS - test_send_receive_ext_id in 0.004 seconds
===================================================================
START - test_send_receive_ext_id_masked
 PASS - test_send_receive_ext_id_masked in 0.004 seconds
===================================================================
START - test_send_receive_ext_id_rtr
 SKIP - test_send_receive_ext_id_rtr in 0.001 seconds
===================================================================
START - test_send_receive_msgq
 PASS - test_send_receive_msgq in 0.010 seconds
===================================================================
START - test_send_receive_std_id
 PASS - test_send_receive_std_id in 0.003 seconds
===================================================================
START - test_send_receive_std_id_masked
 PASS - test_send_receive_std_id_masked in 0.003 seconds
===================================================================
START - test_send_receive_std_id_no_data
 PASS - test_send_receive_std_id_no_data in 0.001 seconds
===================================================================
START - test_send_receive_std_id_rtr
 SKIP - test_send_receive_std_id_rtr in 0.001 seconds
===================================================================
START - test_send_receive_wrong_id
 PASS - test_send_receive_wrong_id in 0.101 seconds
===================================================================
START - test_send_std_id_dlc_of_range
 PASS - test_send_std_id_dlc_of_range in 0.001 seconds
===================================================================
START - test_send_std_id_out_of_range
E: invalid frame with standard (11-bit) CAN ID 0x800
 PASS - test_send_std_id_out_of_range in 0.005 seconds
===================================================================
START - test_send_while_stopped
 PASS - test_send_while_stopped in 0.001 seconds
===================================================================
START - test_set_bitrate
 PASS - test_set_bitrate in 0.001 seconds
===================================================================
START - test_set_bitrate_too_high
 PASS - test_set_bitrate_too_high in 0.001 seconds
===================================================================
START - test_set_bitrate_too_low
 SKIP - test_set_bitrate_too_low in 0.001 seconds
===================================================================
START - test_set_bitrate_while_started
 PASS - test_set_bitrate_while_started in 0.001 seconds
===================================================================
START - test_set_mode_while_started
 PASS - test_set_mode_while_started in 0.001 seconds
===================================================================
START - test_set_state_change_callback
 PASS - test_set_state_change_callback in 0.001 seconds
===================================================================
START - test_set_timing_max
 PASS - test_set_timing_max in 0.001 seconds
===================================================================
START - test_set_timing_min
 PASS - test_set_timing_min in 0.001 seconds
===================================================================
START - test_set_timing_while_started
 PASS - test_set_timing_while_started in 0.001 seconds
===================================================================
START - test_start_while_started
 PASS - test_start_while_started in 0.001 seconds
===================================================================
START - test_stop_while_stopped
 PASS - test_stop_while_stopped in 0.001 seconds
===================================================================
TESTSUITE can_classic succeeded
Running TESTSUITE can_stats
===================================================================
START - test_can_stats_accessors
 PASS - test_can_stats_accessors in 0.001 seconds
===================================================================
TESTSUITE can_stats succeeded
Running TESTSUITE can_timing
===================================================================
testing on device can@2010d000 @ 8000000 Hz, sample point margin +/-50 permille
Warning: Testing limited selection of bitrates (CONFIG_TEST_ALL_BITRATES=n)
START - test_timing
testing bitrate 20000, sample point 87.5%: sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 25 OK, sample point error 0.0%
testing bitrate 50000, sample point 87.5%: sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 10 OK, sample point error 0.0%
testing bitrate 125000, sample point 87.5%: sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 4 OK, sample point error 0.0%
testing bitrate 250000, sample point 87.5%: sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 2 OK, sample point error 0.0%
testing bitrate 500000, sample point 87.5%: sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 1 OK, sample point error 0.0%
testing bitrate 800000, sample point 80.0%: sjw = 1, prop_seg = 0, phase_seg1 = 7, phase_seg2 = 2, prescaler = 1 OK, sample point error 0.0%
testing bitrate 1000000, sample point 75.0%: sjw = 1, prop_seg = 0, phase_seg1 = 5, phase_seg2 = 2, prescaler = 1 OK, sample point error 0.0%
 PASS - test_timing in 0.087 seconds
===================================================================
START - test_timing_data
 SKIP - test_timing_data in 0.001 seconds
===================================================================
TESTSUITE can_timing succeeded
CAN transceiver device not readyRunning TESTSUITE can_utilities
===================================================================
START - test_can_bytes_to_dlc
 PASS - test_can_bytes_to_dlc in 0.001 seconds
===================================================================
START - test_can_dlc_to_bytes
 PASS - test_can_dlc_to_bytes in 0.001 seconds
===================================================================
START - test_can_frame_matches_filter
 PASS - test_can_frame_matches_filter in 0.001 seconds
===================================================================
TESTSUITE can_utilities succeeded
ERROR: Test suite 'can_transceiver' did not run.
```
Initially some tests failed since the driver didn't account for `CONFIG_CAN_ACCEPT_RTR`. This was fixed by:
```diff
diff --git a/drivers/can/can_mchp_mss.c b/drivers/can/can_mchp_mss.c
index 95dca31b727..4f93823886b 100644
--- a/drivers/can/can_mchp_mss.c
+++ b/drivers/can/can_mchp_mss.c
@@ -411,8 +411,6 @@ static int mss_can_add_rx_filter(const struct device *dev, can_rx_callback_t cal
                acr <<= 29 - 11;
        }
 
-       amr |= MSS_CAN_RX_MSG_ACC_RTR;
-
        struct mss_can_data *data = dev->data;
        const struct mss_can_config *cfg = dev->config;
        uintptr_t reg_base = cfg->reg_base;
@@ -795,9 +793,15 @@ static void mss_can_irq_rx_available(const struct device *dev, struct mss_can_da
 
                k_spin_unlock(&data->rx_lock, key);
 
-               if (cb != NULL) {
-                       cb(dev, &frame, user_data);
+#ifndef CONFIG_CAN_ACCEPT_RTR
+               if ((frame.flags & CAN_FRAME_RTR) == 0) {
+#endif
+                       if (cb != NULL) {
+                               cb(dev, &frame, user_data);
+                       }
+#ifndef CONFIG_CAN_ACCEPT_RTR
                }
+#endif
 
                /* We are done with this buffer for now */
                rx_buf_status &= ~(UINT32_C(1) << index);
```
